### PR TITLE
w64wrapper for poly1305

### PR DESCRIPTION
**Pull request by JacobBarthelmeh**
Helps with use of poly1305 in environments that do not have word64 available like the pic18f.

### CODE REVIEW

File: wolfcrypt/src/misc.c [hunk 1]
```diff
@@ -796,6 +796,13 @@ WC_MISC_STATIC WC_INLINE w64wrapper w64ShiftLeft(w64wrapper a, int shift)
     return a;
 }
 
+WC_MISC_STATIC WC_INLINE w64wrapper w64Mul(unsigned int a, unsigned int b)
```
Comment (SparkiDev) [inline wolfcrypt/src/misc.c:799]: word32?

File: wolfcrypt/src/misc.c [hunk 2]
```diff
@@ -841,6 +848,24 @@ WC_MISC_STATIC WC_INLINE w64wrapper w64Add32(w64wrapper a, word32 b, byte *wrap)
     return a;
 }
 
+WC_MISC_STATIC WC_INLINE w64wrapper w64Add(w64wrapper a, w64wrapper b,
+    byte *wrap)
+{
+    a.n[1] = a.n[1] + b.n[1];
```
Comment (SparkiDev) [inline wolfcrypt/src/misc.c:854]: +=

File: wolfcrypt/src/misc.c [hunk 3]
```diff
@@ -841,6 +848,24 @@ WC_MISC_STATIC WC_INLINE w64wrapper w64Add32(w64wrapper a, word32 b, byte *wrap)
...
+    if (a.n[1] < b.n[1]) {
+        a.n[0]++;
+        if (wrap != NULL && a.n[0] == 0)
+                *wrap = 1;
+    }
+
+    a.n[0] = a.n[0] + b.n[0];
```
Comment (SparkiDev) [inline wolfcrypt/src/misc.c:861]: +=

**Review by SparkiDev — CHANGES_REQUESTED** [review]

File: wolfcrypt/src/misc.c [hunk 4]
```diff
@@ -796,6 +806,13 @@ WC_MISC_STATIC WC_INLINE w64wrapper w64ShiftLeft(w64wrapper a, int shift)
     return a;
 }
 
+WC_MISC_STATIC WC_INLINE w64wrapper w64Mul(word32 a, word32 b)
+{
+    w64wrapper ret;
+    ret.n = (word64)a * (word64)b;
```
Comment (SparkiDev) [inline wolfcrypt/src/misc.c:812]: You only need to cast one of them to word64.
```
ret.n = (word64)a * b;
```

File: wolfcrypt/src/misc.c [hunk 5]
```diff
@@ -716,6 +716,16 @@ WC_MISC_STATIC WC_INLINE w64wrapper w64Add32(w64wrapper a, word32 b, byte *wrap)
     return a;
 }
 
+WC_MISC_STATIC WC_INLINE w64wrapper w64Add(w64wrapper a, w64wrapper b,
+    byte *wrap)
+{
+    a.n = a.n + b.n;
```
Comment (SparkiDev) [inline wolfcrypt/src/misc.c:722]: +=
(And above in w64Add32)

File: wolfcrypt/src/misc.c [hunk 6]
```diff
@@ -841,6 +858,24 @@ WC_MISC_STATIC WC_INLINE w64wrapper w64Add32(w64wrapper a, word32 b, byte *wrap)
...
+{
+    a.n[1] += b.n[1];
+    if (a.n[1] < b.n[1]) {
+        a.n[0]++;
+        if (wrap != NULL && a.n[0] == 0)
+                *wrap = 1;
+    }
+
+    a.n[0] += b.n[0];
+    if (a.n[0] < b.n[0]) {
```
Comment (SparkiDev) [inline wolfcrypt/src/misc.c:872]: Chack warp != NULL

File: wolfcrypt/src/misc.c [hunk 7]
```diff
@@ -961,6 +996,30 @@ WC_MISC_STATIC WC_INLINE w64wrapper w64ShiftLeft(w64wrapper a, int shift)
...
+    w64wrapper ret;
+    word16 ltlA, ltlB, ltlC, ltlD;
+    word32 bigA, bigB, bigC, bigD;
+
+    ltlA = a & 0xFFFF;
+    ltlB = (a >> 16) & 0xFFFF;
+    ltlC = b & 0xFFFF;
+    ltlD = (b >> 16) & 0xFFFF;
+
+    bigA = (word32)ltlA * (word32)ltlC;
```
Comment (SparkiDev) [inline wolfcrypt/src/misc.c:1010]: Don't need to cast both to word32.

**Review by SparkiDev — CHANGES_REQUESTED** [review]

File: wolfcrypt/src/misc.c (hunk 4 above)
Comment (JacobBarthelmeh) [reply wolfcrypt/src/misc.c:812]: would rather be explicit than rely on the compiler, especially the old compiler that was being used with the pic18f work

**Review by SparkiDev — APPROVED** [review]

**Pull request merged**